src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs

```diff
@@ -211,36 +175,51 @@ public int TakeData(byte[] buffer, int dataOffset, int size)
                 size = _length - _offset;
             }
 
-            Buffer.BlockCopy(_data, _offset, buffer, dataOffset, size);
+            Buffer.BlockCopy(_data, _header + _offset, buffer, dataOffset, size);
             _offset += size;
             return size;
         }
 
+
+        /// <summary>
+        /// Set the MARS SMUX header information for this packet
+        /// </summary>
+        /// <param name="header">the header to write into the packet</param>
+        public void SetHeader(SNISMUXHeader header)
+        {
+            Debug.Assert(header != null, "writing null mux header to packet");
+
+            Debug.Assert(_offset == 0, "writing mux header to partially read packet");
+            Debug.Assert(_header == SNISMUXHeader.HEADER_LENGTH, "writing mux header to partially incorrectly sized reserved region");
+            Debug.Assert(((_flags & SNIPacketFlags.MuxHeaderReserved) == SNIPacketFlags.MuxHeaderReserved), "writing mux heaser to non-mux packet");
+
+            header.Write(_data.AsSpan(0, SNISMUXHeader.HEADER_LENGTH));
+            _capacity += _header;
```

Comment:
Why is the _capacity being incremented by the size of the _header? Sounds like you are incrementing this to set the new offset of capacity? 
Basically the name of the member int and what it is doing seem out of sync

Reply:
It seems like when the SMUX header is set, you are telling this packet that now your capacity is original capacity + the header, that is written by MUX and hence the length will increase as well.

Reply:
Once you allocate the header it's part of the data that needs to be sent so it's length is added to the total data. I chose the model where the header is a reserved space but once used it's part of the packet, it made thinking about the muxheader as a part of the packet easier. The renaming has made this feel clearer to me, see what you think.